src/analyze_streams.cpp

```diff
@@ -0,0 +1,91 @@
+#include <migraphx/analyze_streams.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool happens_before(const std::vector<std::size_t>& e1, const std::vector<std::size_t>& e2)
+{
+    return std::equal(e1.begin(), e1.end(), e2.begin(), e2.end(), std::less_equal<>{}) and
+           not std::equal(e1.begin(), e1.end(), e2.begin(), e2.end(), std::greater_equal<>{});
```

Comment:
If I understand this correctly, if some or all values in e1 are less than e2, then e1 happens before e2. However, if all of the values are equal, e1 does not happen before e2. I am not sure why `greater_equal` is needed since the first predicate for `less_equal` would return false if any value is greater. I think the second predicate can be just `equal`.

Reply:
The logic is that all of them should be less than equal and at least one of them is strictly less than. See [here](https://en.wikipedia.org/wiki/Vector_clock#Partial_ordering_property).

Reply:
Both logic cases will work, but it won't be necessary to check `greater` in this case. You can change it to just `std::equal` for the second conditional or leave it otherwise.